src/optimizer/limit_push_down_optimizer.cpp

```diff
@@ -0,0 +1,136 @@
+    plan->setLastOperator(visitOperator(plan->getLastOperator()));
+}
+
+static LogicalScanNodeTable* isSimpleScanNode(const std::shared_ptr<LogicalOperator>& op) {
+    switch (op->getOperatorType()) {
+    case LogicalOperatorType::SCAN_NODE_TABLE: {
+        return &op->cast<LogicalScanNodeTable>();
+    }
+    case LogicalOperatorType::PROJECTION: {
+        return isSimpleScanNode(op->getChild(0));
+    }
+    default:
+        return nullptr;
+    }
+}
+
+std::shared_ptr<LogicalOperator> LimitPushDownOptimizer::finishPushDown(
+    std::shared_ptr<LogicalOperator> op) {
+    if (limitOperator == nullptr) {
+        return op;
+    }
+    LogicalPlan plan;
+    plan.setLastOperator(op);
+    planner::Planner planner(context);
+    planner.appendMultiplicityReducer(plan);
+    planner.appendLimit(limitOperator->getSkipNum(), limitOperator->getLimitNum(), plan);
+    limitOperator = nullptr;
+    return plan.getLastOperator();
+}
+
+std::shared_ptr<LogicalOperator> LimitPushDownOptimizer::visitOperator(
+    std::shared_ptr<LogicalOperator> op) {
+    switch (op->getOperatorType()) {
+    case LogicalOperatorType::LIMIT: {
+        KU_ASSERT(this->limitOperator == nullptr);
+        this->limitOperator = &op->cast<LogicalLimit>();
+        KU_ASSERT(op->getChild(0)->getOperatorType() == LogicalOperatorType::MULTIPLICITY_REDUCER);
+        // Remove limit
+        return visitOperator(op->getChild(0)->getChild(0));
+    }
+    case LogicalOperatorType::EXPLAIN:
+    case LogicalOperatorType::SEMI_MASKER:
+    case LogicalOperatorType::ACCUMULATE:
+    case LogicalOperatorType::PATH_PROPERTY_PROBE:
+    case LogicalOperatorType::PROJECTION: {
+        op->setChild(0, visitOperator(op->getChild(0)));
+        return op;
+    }
+    case LogicalOperatorType::DISTINCT: {
+        if (limitOperator == nullptr) {
+            return op;
+        }
+        // AGGREGATE's limit is task-level restrictions
+        auto& distinctOp = op->cast<LogicalDistinct>();
+        distinctOp.setLimitNum(limitOperator->getLimitNum());
+        distinctOp.setSkipNum(limitOperator->getSkipNum());
+        // We can't remove this limit, because there can be multiple tasks performing AGGREGATE
+        return finishPushDown(op);
+    }
+    case LogicalOperatorType::UNION_ALL: {
+        for (auto i = 0u; i < op->getNumChildren(); ++i) {
+            auto optimizer = LimitPushDownOptimizer(context);
+            op->setChild(i, optimizer.visitOperator(op->getChild(i)));
+        }
+        return op;
+    }
+    case LogicalOperatorType::HASH_JOIN: {
+        if (limitOperator == nullptr) {
+            return op;
+        }
+        auto& hashJoin = op->cast<LogicalHashJoin>();
+        // Push down limit to left side
+        if (hashJoin.getJoinType() == JoinType::LEFT) {
+            // Add a limit to the probe side,and keep the original limit as well.
+            auto optimizer = LimitPushDownOptimizer(context);
```

Comment:
Is there a specific benefit by pushing LIMIT to probe side? Hash join probe is a pipelined operator instead of a  sink operator, so the improvement is probably not significant in my opinion?

Reply:
The execution plan on the probe side may also be split into multiple stages, so after pushing down, the limit can be pushed down to the downstream stage.